tests/testthat/test-filter.R

```diff
@@ -0,0 +1,100 @@
+# TODO: implement these functions in filters.R and use them in this test
+include <- function(...) {}
+exclude <- function(...) {}
+lessthan <- function(...) {}
+morethan <- function(...) {}
+interval <- function(...) {}
+
+# TODO: implement filters, check number of samples after filtering the test data and fill expected_samples
+test_cases <- list(
+  list(
+    method = "by.meta", query = list(Status = include("C")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.meta", method = "by.meta", query = list(Lane = exclude("A")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.meta", query = list(Age = lessthan(23)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.meta", query = list(Age = interval(15, 23)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.meta", query = list(Lane = include("B", "C")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.meta", query = list(Lane = exclude("A", "C")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.meta", query = list(Lane = include("A", "B"), Age = morethan(15)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.repertoire", query = list(n_clonotypes = morethan(1000)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.repertoire", query = list(n_clones = lessthan(100)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.rep", query = list(n_clonotypes = morethan(1000)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.rep", query = list(n_clones = lessthan(100)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.clonotype", query = list(CDR3.aa = exclude("partial", "out_of_frame")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.clonotype", query = list(Clones = interval(5, 50)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.cl", query = list(CDR3.aa = exclude("partial", "out_of_frame")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.cl", query = list(Clones = interval(5, 50)),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.gene", query = list(V = exclude("TRBV1", "TRGV11")),
+    expected_samples = 0
+  ),
+  list(
+    method = "by.gene", query = list(V = include("TRAV9"), J = include("TRAJ11")),
+    expected_samples = 0
```

Comment:
When you test only filter all case, you do not test all the other scenarios. Try using example datasets from immunarch to create positive tests. Otherwise any filtering function can always return zero-length result, which is not the desired functionality

Reply:
Agree, I expect that there will be tests with value of samples/clonotypes after manual filtering as expected samples/clonotypes; without this, we cannot check all works properly when you implement this filters

Reply:
Counted expected numbers manually for default test dataset